### DISCUSSION TITLE
USB only REPL

### DISCUSSION THREAD

**Discussion by ghost** [micropython/micropython#11305] [ESP32]
Hi everyone,
I'm working on an ESP32-S3 and I'm trying to move the REPL from the UART0 to the USB only. I've tried to chande the sdkconfig file but I always get the REPL on both UART0 and USB also if the compile flags are for USB only.

I just want to know if it's possible to use REPL on USB and leave the 3 UART free in order to be used as ordinary UART ports.

I've also tried the os.dupterm(None,0) command but nothing happens.

Thank you for your help.
Greetings,
Andrea

**Comment by IlysvlVEizbr** [comment]
I have the same issue, and would like to use USB + 3 free UART.

**Comment by ghost** [comment]
At the moment I've changed the following define into _mpconfigboard.h_

`#define MICROPY_HW_ENABLE_UART_REPL         (0)`

and the REPL doesn't outputs on UART0, I'm able to see the USB REPL only.
I configure the UART0 as follows:

`uart=UART(0,baudrate=115200,tx=43,rx=44)`

and tried to write something on the UART but nothing happens. There is no output on UART0.

**Comment by ThinkTransit** [comment]
I have looked at this before and I think the problem is that it is not possible to change the pins of uart0. See here, it looks like they are hard coded.

https://github.com/micropython/micropython/blob/9e6885ad820a9f42bf93df56eb3f3be0c8639622/ports/esp32/machine_uart.c#L338

**Reply by IlysvlVEizbr** [reply]
Yes, UART0 uses a fixed GPIO pin which is determined by the ESP32S3 hardware itself, but at the same time, the USB function and pins (GPIO19, 20) on ESP32S3 can be separated from UART0 (when not using Micropython official firmware). So I wonder to know if it is possible to separate the two function in the official firmware and redirect the REPL function to the USB port, so that we can release UART0 resource.

**Reply by ghost** [reply]
Thanks @IlysvlVEizbr , that's what I mean. It's not a problem having fixed GPIO on UART0. The problem here is that we have two possibilities:
1) Use the REPL on both USB and UART0
2) not using REPL on UART 0 ( defining MICROPY_HW_ENABLE_UART_REPL to 0) but UART0 cannot be accessed and used as ordinary UART

**Comment by straga** [comment]
ESP32S2 not works too.
UART1 - work
UART0 - not work


>>> import machine
>>> uart = machine.UART(0, baudrate=9600, tx=12, rx=13)
>>> uart.write("3434")
>>> uart.read()
>>> uart.read(1024)
>>>
>>>
>>>
>>> uart1 = machine.UART(1, baudrate=9600, tx=4, rx=5)
>>> uart1.write("3454")
>>> 4
>>> uart1.read(1024)
>>> b'3454'

**Reply by straga** [reply]
Probably same: https://github.com/orgs/micropython/discussions/10505
